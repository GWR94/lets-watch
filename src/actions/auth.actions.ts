import { LOGIN, LOGOUT, LoginAction, LogoutAction } from "../interfaces/auth.redux.i";

export const login = (uid: string): LoginAction => ({
  type: LOGIN,
  uid,
});

export const logout = (): LogoutAction => ({
  type: LOGOUT,
});
