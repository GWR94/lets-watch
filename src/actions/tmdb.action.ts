import axios from "axios";
import { Dispatch } from "redux";
import {
  FETCH_TOP_MOVIES,
  FETCH_TOP_TV,
  APIResponse,
} from "../interfaces/tmdb.actions.i";

export const fetchTopMovies = (): ((dispatch: Dispatch) => Promise<void>) => async (
  dispatch,
): Promise<void> => {
  const res = await axios.get(
    `${process.env.TMDB_URL}/movie/popular${process.env.TMDB_API_KEY}`,
  );
  dispatch({
    type: FETCH_TOP_MOVIES,
    payload: res.data.results,
  });
};

export const fetchTopTV = (): ((dispatch: Dispatch) => Promise<void>) => async (
  dispatch,
): Promise<void> => {
  const res = await axios.get(
    `${process.env.TMDB_URL}/tv/popular${process.env.TMDB_API_KEY}`,
  );
  dispatch({
    type: FETCH_TOP_TV,
    payload: res.data.results,
  });
};
