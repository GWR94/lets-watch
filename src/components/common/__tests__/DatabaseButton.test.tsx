import { mount } from "enzyme";
import React from "react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import axios from "axios";
import * as actions from "../../../actions/auth.actions";
import { ADD_TO_USER } from "../../../interfaces/auth.redux.i";

describe("Database Button test cases", () => {
  let wrapper;
  const setState = jest.fn();
  const useStateSpy: any = jest.spyOn(React, "useState");
  useStateSpy.mockImplementation((init) => [init, setState]);

  afterEach(() => {
    jest.clearAllMocks();
  });

  const mockStore = configureStore([thunk]);
  const initialState = {
    auth: {
      profile: {
        tvShows: [12, 245, 656],
        movies: [2, 155, 222],
      },
    },
  };
  let store;
  jest.mock("react-redux");
  jest.mock("axios");

  beforeEach(() => {
    store = mockStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <DatabaseButton type="movie" id={123} />
      </Provider>,
    );
  });

  describe("snapshot test", () => {
    it("should render the connected component", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("store based test cases", () => {
    it("should return the correct button text when the id is not in store for movies", () => {
      const buttonText = wrapper.find(".btn").text();
      expect(buttonText).toBe("Add to My Movies");
    });
    it("should return the correct button text when the id is not in store for tv shows", () => {
      wrapper = mount(
        <Provider store={store}>
          <DatabaseButton type="tv" id={123} />
        </Provider>,
      );
      const buttonText = wrapper.find(".btn").text();
      expect(buttonText).toBe("Add to My TV Shows");
    });
    it("should return the correct button text when the id is in the store for movies", () => {
      wrapper = mount(
        <Provider store={store}>
          <DatabaseButton type="movie" id={2} />
        </Provider>,
      );
      const buttonText = wrapper.find(".btn").text();
      expect(buttonText).toBe("Remove from My Movies");
    });
    it("should return the correct button text when the id is in the store for tv shows", () => {
      wrapper = mount(
        <Provider store={store}>
          <DatabaseButton type="tv" id={12} />
        </Provider>,
      );
      const buttonText = wrapper.find(".btn").text();
      expect(buttonText).toBe("Remove from My TV Shows");
    });
    it("should dispatch the correct action when the button is clicked when not in store", async () => {
      wrapper
        .find(".modal__button")
        .at(0)
        .simulate("click");
      expect(setState).toBeCalledWith(true);
      await store.dispatch(actions.addToUser(123, "movies"));
      expect(store).toEqual({
        auth: {
          profile: {
            tvShows: [12, 245, 656],
            movies: [2, 155, 222],
          },
        },
      });
    });
  });
});
