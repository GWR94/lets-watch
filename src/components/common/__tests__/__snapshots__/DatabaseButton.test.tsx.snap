// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Database Button test cases snapshot test should render the connected component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <DatabaseButton
    id={123}
    type="movie"
  >
    <div
      className="modal__button-container"
    >
      <Button
        className="modal__button"
        color="success"
        onClick={[Function]}
        outline={true}
        size="lg"
        style={
          Object {
            "display": "inline-block",
          }
        }
        tag="button"
      >
        <button
          aria-label={null}
          className="modal__button btn btn-outline-success btn-lg"
          onClick={[Function]}
          style={
            Object {
              "display": "inline-block",
            }
          }
          type="button"
        >
          Add to My Movies
        </button>
      </Button>
    </div>
  </DatabaseButton>
</Provider>
`;
